bigchaindb/core.py

```diff
@@ -555,44 +560,44 @@ def write_vote(self, block, vote, block_number):
         if self.has_previous_vote(block):
             return None
 
-        update = {'votes': r.row['votes'].append(vote)}
-
         # We need to *not* override the existing block_number, if any
         # FIXME: MIGHT HAVE RACE CONDITIONS WITH THE OTHER NODES IN THE FEDERATION
-        if 'block_number' not in block:
-            update['block_number'] = block_number
+        if 'block_number' not in vote:
+            vote['block_number'] = block_number  # maybe this should be in the signed part...or better yet, removed..
 
-        r.table('bigchain') \
-            .get(vote['vote']['voting_for_block']) \
-            .update(update) \
+        r.table('votes') \
+            .insert(vote) \
             .run(self.conn)
 
-    def get_last_voted_block(self):
+    def get_last_voted_block(self, return_block_number=False):
```

Comment:
That's the problem, blocks never have block numbers...they're stored in the votes themselves.  So I made this kind of ugly compromise to avoid rewriting the voter module (if you want the block number, that's an additional database read -- but as you said, this is the only function that uses it).

Yet another reason to get rid of them, right?